pkg/k8s/identitybackend/identity.go

```diff
@@ -0,0 +1,418 @@
+	}
+
+	for _, identityObject := range c.Store.List() {
+		identity, ok := identityObject.(*types.Identity)
+		if !ok {
+			return nil
+		}
+
+		if reflect.DeepEqual(identity.Labels, labels) {
+			return identity
+		}
+	}
+
+	return nil
+}
+
+// Get returns the ID which is allocated to a key in the identity CRDs in
+// kubernetes.
+// Note: the lock field is not supported with the k8s CRD allocator.
+func (c *crdBackend) Get(ctx context.Context, key allocator.AllocatorKey) (idpool.ID, error) {
+	identity := c.get(ctx, key)
+	if identity == nil {
+		return idpool.NoID, nil
+	}
+
+	id, err := strconv.ParseUint(identity.Name, 10, 64)
+	if err != nil {
+		return idpool.NoID, fmt.Errorf("unable to parse value '%s': %s", identity.Name, err)
+	}
+
+	return idpool.ID(id), nil
+}
+
+func (c *crdBackend) GetIfLocked(ctx context.Context, key allocator.AllocatorKey, lock kvstore.KVLocker) (idpool.ID, error) {
+	return c.Get(ctx, key)
+}
+
+// GetByID returns the key associated with an ID. Returns nil if no key is
+// associated with the ID.
+// Note: the lock field is not supported with the k8s CRD allocator.
+func (c *crdBackend) GetByID(id idpool.ID) (allocator.AllocatorKey, error) {
+	if c.Store == nil {
+		return nil, fmt.Errorf("store is not available yet")
+	}
+
+	identityTemplate := &types.Identity{
+		CiliumIdentity: &v2.CiliumIdentity{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: id.String(),
+			},
+		},
+	}
+
+	obj, exists, err := c.Store.Get(identityTemplate)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, nil
+	}
+
+	identity, ok := obj.(*types.Identity)
+	if !ok {
+		return nil, fmt.Errorf("invalid object")
+	}
+
+	return c.KeyType.PutKeyFromMap(identity.Labels), nil
+}
+
+// Release dissociates this node from using the identity bound to key. When an
+// identity has no references it may be garbage collected.
+func (c *crdBackend) Release(ctx context.Context, key allocator.AllocatorKey) (err error) {
+	identity := c.get(ctx, key)
+	if identity == nil {
+		return fmt.Errorf("unable to release identity %s: identity does not exist", key)
+	}
+
+	if _, ok := identity.Status.Nodes[c.NodeName]; !ok {
+		return fmt.Errorf("unable to release identity %s: identity is unused", key)
+	}
+
+	delete(identity.Status.Nodes, c.NodeName)
+
+	capabilities := k8sversion.Capabilities()
+
+	identityOps := c.Client.CiliumV2().CiliumIdentities()
+	if capabilities.Patch {
+		var patch []byte
+		patch, err = json.Marshal([]JSONPatch{
+			{
+				OP:   "delete",
+				Path: "/status/nodes/" + c.NodeName,
+			},
+		})
+		if err != nil {
+			return err
+		}
+		_, err = identityOps.Patch(identity.GetName(), k8sTypes.JSONPatchType, patch, "status")
+		if err == nil {
+			return nil
+		}
+		/* fall through and attempt UpdateStatus() or Update() */
+	}
+
+	identityCopy := identity.DeepCopy()
+	if identityCopy.Status.Nodes == nil {
+		return nil
+	}
+
+	if capabilities.UpdateStatus {
+		_, err = identityOps.UpdateStatus(identityCopy.CiliumIdentity)
```

Comment:
Do we need retries here? If Update fails because of conflicts, we will never release the identity.

Reply:
pkg/allocator.Allocator retries, but only for allocate. This includes calls AcquireReference. In theory, a failed Release will be garbage collected eventually. In any case, the kvstore backend also doesn't retry at that layer. Are you ok with slightly different behaviours between the backends?

Reply:
Discussed online. I'll add some more comments/logging to ensure this case is clear, but pkg/allocator is the primary retry source.